torch/csrc/jit/tensorexpr/registerizer.h

```diff
@@ -0,0 +1,354 @@
+      // already a scalar.
+      return;
+    }
+
+    SimplifierHashType accessHash = hasher_.hash(v->buf());
+    for (auto* i : v->indices()) {
+      accessHash = hasher_.hash_combine(accessHash, i);
+    }
+    accessHash = hasher_.hash_combine(accessHash, v->mask());
+
+    auto candidateIt = candidates_.find(accessHash);
+    if (candidateIt == candidates_.end()) {
+      candidates_[accessHash] =
+          new AccessInfo(v->buf(), v->indices(), enclosingBlock_, loopCost_, v);
+      return;
+    }
+
+    AccessInfo* info = candidateIt->second;
+    info->addLoad(v, enclosingBlock_, loopCost_);
+  }
+
+  std::vector<AccessInfo*> getCandidates() {
+    std::vector<AccessInfo*> ret;
+    std::unordered_map<const Buf*, std::vector<AccessInfo*>> access_by_buf;
+    for (const auto& pair : candidates_) {
+      AccessInfo* info = pair.second;
+      // Now that analysis is complete we must normalize the costs by the parent
+      // Block we plan to insert the scalar var into.
+      info->store_cost = IRSimplifier::simplify(
+          new Div(info->store_cost, costByBlock_[info->parent]));
+
+      info->load_cost = IRSimplifier::simplify(
+          new Div(info->load_cost, costByBlock_[info->parent]));
+
+      access_by_buf[pair.second->buf].push_back(pair.second);
+    }
+
+    for (const auto& pair : access_by_buf) {
+      const Buf* buf = pair.first;
+      const std::vector<AccessInfo*>& accesses = pair.second;
+      for (auto* info : accesses) {
+        // Filter out low cost accesses.
+        if (info->store_cost->isConstant() &&
+            immediateAs<int>(info->store_cost) <= 1) {
+          continue;
+        }
+
+        bool any_overlap = false;
+        // We want to find possible overlaps from other accesses.
+        for (auto* i2 : accesses) {
+          if (info == i2) {
+            continue;
+          }
+
+          // All accesses to a buf must have the same dimensionality.
+          assert(info->indices.size() == i2->indices.size());
+
+          // They don't overlap if there is a guaranteed difference in any
+          // dimension.
+          bool overlap = true;
+          for (size_t i = 0; i < info->indices.size(); ++i) {
+            const Expr* diff = new Sub(info->indices[i], i2->indices[i]);
+            diff = IRSimplifier::simplify(diff);
+            if (diff->isConstant() && !immediateEquals(diff, 0)) {
+              overlap = false;
+              break;
+            }
+          }
+
+          if (overlap) {
+            any_overlap = true;
+            break;
+          }
+        }
+        if (!any_overlap) {
+          ret.push_back(info);
+        }
+      }
+    }
+    return ret;
+  }
+
+  std::unordered_map<SimplifierHashType, AccessInfo*> candidates_;
+  std::unordered_map<const Block*, const Expr*> costByBlock_;
+
+  const Expr* loopCost_;
+  const Block* enclosingBlock_;
+  HashProvider hasher_;
+};
+
+// Walks the IR an replaces a single Acccess with a local scalar Var.
```

Comment:
Non-Blocking: I am perfectly fine with this approach in general. But I would like to point out there are a lot more cases to make it functionally correct in slightly more general cases. 

From a certain perspective, the Registerizer move a global memory access to a thread-local memory. This could change the semantics if the memory access has cross-thread dependency. For example: if a global reads really needs to read the information of another atomic global writes, that access really needs to get through the global memory. 

This is not likely a problem because we cannot generate that complex a program yet. But we should keep reminding ourselves of the memory semantics change.

Reply:
Right, this only addresses a subset of cases where you can push accesses to a scalar. I believe it's currently pessimistic, however, if there are any accesses which may overlap a registerization candidate program-wide we won't do it. So we should be correct always but we'll leave some perf on the table.

The next step is to divide the program into sub-sections which can have distinct registerizations and write them back at the boundaries of those subsections. I have some ideas on that, but it gets more complicated.